Hw 5 chunck allocator

byte-yu opened the pull request:
Домашняя работа номер 5

Costello1329 committed:
added new task

Costello1329 committed:
added video description

Costello1329 committed:
Update readme.md

byte-yu committed:
initial commit

byte-yu committed:
removed cout

byte-yu committed:
del copy for Chuck

byte-yu committed:
renamed and template class written in one file

SweetWeakness commented on chuck_allocator/chunk_allocator.h:
```diff
@@ -0,0 +1,286 @@
+        size_t size_of_data; // size of block part of spent memory
+    public:
+        const static size_t CHUNK_SIZE = 100; //bytes
+        Chunk *prev;
```
Нет, на предыдущие указатель не надо соxранять. Нужно на следующий, иначе работать верно не будет. Подумайте о том что будет происxодить при аллокации нового чанка ПОСЛЕ ситуации a = b

SweetWeakness commented on chuck_allocator/chunk_allocator.h:
```diff
@@ -0,0 +1,286 @@
+        Chunk *chunks = nullptr; // chunks
+
+        void report(T *p, size_type n, bool alloc = true) {
+            std::cout << (alloc ? "Alloc: " : "Dealloc: ") << sizeof(T) * n
```
#define DEBUG ?

SweetWeakness commented on chuck_allocator/chunk_allocator.h:
```diff
@@ -0,0 +1,286 @@
+
+    template<typename T>
+    class MyAllocator {
+        static uint copies;
```
Нет, это не правильно. Никакиx static-полей быть не должно. Подумайте почему

SweetWeakness commented on chuck_allocator/chunk_allocator.h:
```diff
@@ -0,0 +1,286 @@
+
+        MyAllocator<T> &operator=(MyAllocator<T> const &other);
+
+        T *allocate(size_t n);
```
У вас же есть pointer, size_type. Зачем вам T*, size_t ???

SweetWeakness reviewed (changes requested):
PS: Это Константин, добавьте меня контрибом в репу! Скиньте инвайт в телеграм

byte-yu committed:
Simple List + Node

byte-yu committed:
Allocator :: allocate

byte-yu committed:
chunk skip copy-constructor

byte-yu committed:
test with copy constructor and copy operator

byte-yu committed:
default copy constructor and copy operator
TODO: destructor with multiple allocators

byte-yu committed:
additional test for copy added

byte-yu committed:
self_counter added

byte-yu committed:
test for vector

byte-yu committed:
address of list added in self_report function

byte-yu committed:
DEBUG: added address of the list to copy-operator
Allocator: initial pointer of list is nullptr

devyueightfive commented:
1. Перестроил структуру по парадигме Allocator->List->Node->Chunk
2. List хранит только начало
3. Счетчик строится на куче. В конструкторе копирования присваиваются ссылки на List и счетчик (+увеличивается). В операторе копирования проверяется предыдущий счетчик. В случае единственности ссылки аллокатора на List, вызывается деструкторы List и счетчика. 
4. Используется DEBUG для отладочных записей
5. от T, T*, size_t перешел к value_type, pointer, size_type

byte-yu committed:
test copy of unallocated allocator

byte-yu committed:
removed comments

byte-yu committed:
remove cmake
